---
title: "Using heatmaply with biological data (mutational and gene expression data)"
date: "`r Sys.Date()`"
author: "Alan O'Callaghan"
output:
  html_document:
    self_contained: yes
    toc: true
    fig_width: 15
    fig_height: 15
    depth: 3  # upto three depths of headings (specified by #, ## and ###)  
    number_sections: true  ## if you want number sections at each table header
    theme: yeti  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using heatmaply with biological data}
-->


```{r, echo = FALSE, message = FALSE}
library(heatmaply)
library(heatmaplyExamples)
library(knitr)
knitr::opts_chunk$set(
   # cache = TRUE,
   dpi = 60,
  comment = '#>',
  tidy = FALSE)

```
**Author**: Alan O'Callaghan (alan.b.ocallaghan@gmail.com)




Introduction
============

This vignette is intended to provide an overview of the use of
`heatmaply` in the analysis of biological data. In particular, 
this vignette deals with its use to visualise gene expression 
patterns and sample relationships in 
RNAseq data relating to breast cancer samples,
as retrieved from from [Genomic Data Commons].

Briefly, the following steps were performed:

1. Data retrieval pre-processing
2. Visualisation of raw and voom-transformed data (all genes)
3. Visualisation of raw and voom-transformed data (PAM50 genes only)
4. Visualisation of raw and voom-transformed data 
  (median-centred data, PAM50 genes only)
5. Demonstrate outlier detection using synthetic example

[Genomics Data Commons]: https://gdc.cancer.gov/


Data retrieval and pre-processing
=================
The data used in this vignette is available within the package. 
A full explanation of how this data was retrieved from the 
[Genomic Data Commons] is available in the [data preprocessing vignette].


[Genomics Data Commons]: https://gdc.cancer.gov/
[data preprocessing vignette]: data_preprocessing.html


Visualisation of raw and voom-transformed data
==================

```{r}
cor_mat_raw_logged <- cor(log2(raw_expression + 0.5))

heatmaply(cor_mat_raw_logged, 
    row_side_colors = tcga_brca_clinical,
    main = 'log2 Count data correlation',
    showticklabels = c(FALSE, FALSE),
    subplot_widths = c(0.7, 0.1, 0.2),
    plot_method = 'plotly')

```
voom explanation

```{r}

cor_mat_voomed <- cor(voomed_expression)

heatmaply(cor_mat_voomed, 
    row_side_colors = tcga_brca_clinical,
    main = 'log2 cpm data correlation',
    showticklabels = c(FALSE, FALSE),
    subplot_widths = c(0.7, 0.1, 0.2),
    plot_method = 'plotly')

```



Visualisation of raw data (median-centred data, PAM50 genes only)
==================
Heatmaps of non-centred are shown in [another vignette](non_centred_heatmaps.html) within this package.


```{r}
pam50_genes <- intersect(pam50_genes, rownames(raw_expression))
raw_pam50_expression <- raw_expression[pam50_genes, ]
voomed_pam50_expression <- voomed_expression[pam50_genes, ]

center_raw_mat <- cor_mat_raw_logged - 
    apply(cor_mat_raw_logged, 1, median)

raw_max <- max(abs(center_raw_mat), na.rm=TRUE)
raw_limits <- c(-raw_max, raw_max)


heatmaply(t(center_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    fontsize_col = 7.5,
    col = gplots::bluered(50),
    main = 'raw centered pam50',
    limits = raw_limits,
    plot_method = 'plotly')


heatmaply_cor(cor(center_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'correlation of raw centered pam50',
    plot_method = 'plotly')


```

Visualisation of raw and voom-transformed data (median-centred data, PAM50 genes only)
==================


```{r}
center_voom_mat <- voomed_pam50_expression - 
    apply(voomed_pam50_expression, 1, median)

voom_max <- max(abs(center_voom_mat))
voom_limits <- c(-voom_max, voom_max)


heatmaply(t(center_voom_mat), 
    row_side_colors=tcga_brca_clinical,
    fontsize_col = 7.5,
    showticklabels = c(TRUE, FALSE),
    col = gplots::bluered(50),
    limits = voom_limits,
    main = 'voomed pam50',
    plot_method = 'plotly')


heatmaply_cor(cor(center_voom_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'correlation of voomed pam50',
    plot_method = 'plotly')

```


Demonstrate outlier detection using synthetic example
=====================================================
