---
title: "Using heatmaply with biological data (mutational and gene expression data)"
date: "`r Sys.Date()`"
author: "Alan O'Callaghan"
output:
  html_document:
    self_contained: yes
    toc: true
    fig_width: 15
    fig_height: 15
    depth: 3  # upto three depths of headings (specified by #, ## and ###)  
    number_sections: true  ## if you want number sections at each table header
    theme: yeti  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using heatmaply with biological data}
-->


```{r, echo = FALSE, message = FALSE}
library(heatmaply)
library(heatmaplyExamples)
library(knitr)
knitr::opts_chunk$set(
   # cache = TRUE,
   dpi = 60,
  comment = '#>',
  tidy = FALSE)

```
**Author**: Alan O'Callaghan (alan.b.ocallaghan@gmail.com)




Introduction
============

This vignette is intended to provide an overview of the use of
`heatmaply` in the analysis of biological data. In particular, 
this vignette deals with its use to visualise gene expression 
patterns and sample relationships in 
RNAseq data relating to breast cancer samples,
as retrieved from from [Genomic Data Commons].

Briefly, the following steps were performed:

1. Data retrieval pre-processing
2. Visualisation of raw and voom-transformed data (all genes)
3. Visualisation of raw and voom-transformed data (PAM50 genes only)
4. Visualisation of raw and voom-transformed data 
  (median-centred data, PAM50 genes only)


Data retrieval and pre-processing
=================================
The data used in this vignette is available within the package. 
A full explanation of how this data was retrieved from the 
[Genomic Data Commons] is available in the [data preprocessing vignette].




Visualisation of raw and voom-transformed data
==============================================
In the first instance, one can visualise the pairwise correlation matrix of 
all samples in the matrix of gene counts present in this study (saved
within the package as raw_tcga_breast_expression)





```{r}
cor_mat_raw_logged <- cor(log2(tcga_breast_expression + 0.5))

heatmaply(cor_mat_raw_logged, 
    row_side_colors = tcga_brca_clinical,
    main = 'Correlation of log2 count data (all genes)',
    showticklabels = c(FALSE, FALSE),
    subplot_widths = c(0.7, 0.1, 0.2),
    plot_method = 'plotly')

```


[voom] is part of the limma package. It calculates log2 counts per million reads,
as well as, in this case, performing trimmed mean of M-component normalisation.
voom also calculates precision weights which can be used in linear models. However,
this is outside the scope of this vignette.

[voom]: http://www.statsci.org/smyth/pubs/VoomPreprint.pdf







```{r}


dge <- edgeR::DGEList(counts = tcga_breast_expression)
dge <- edgeR::calcNormFactors(dge, method = "TMM")
voomed_tcga_expression <- limma::voom(dge, normalize.method="quantile")


voomed_tcga_expression <- limma::voom(tcga_breast_expression)
voomed_tcga_expression <- voomed_tcga_expression$E

cor_mat_voomed <- cor(voomed_tcga_expression)

heatmaply(cor_mat_voomed,
    row_side_colors = tcga_brca_clinical,
    main = 'Correlation of normalised data (all genes)',
    showticklabels = c(FALSE, FALSE),
    subplot_widths = c(0.7, 0.1, 0.2),
    plot_method = 'plotly')

```



Visualisation of raw data (median-centred data, PAM50 genes only)
=================================================================
For completeness, heatmaps of non-centred are shown in 
[another vignette](non_centred_heatmaps.html) within this package. 
These can be expected to be similar to the heatmaps shown hereafter, 
but are more difficult to interpret and as a result will not be 
discussed further.


Visualisation of raw and voom-transformed data (median-centred data, PAM50 genes only)
======================================================================================




```{r}
pam50_genes <- intersect(pam50_genes, rownames(tcga_breast_expression))
raw_pam50_expression <- tcga_breast_expression[pam50_genes, ]
voomed_pam50_expression <- voomed_tcga_expression[pam50_genes, ]


log_mat <- log2(raw_pam50_expression + 0.5)
center_raw_mat <- log_mat - 
    apply(log_mat, 1, median)

raw_max <- max(abs(center_raw_mat), na.rm=TRUE)
raw_limits <- c(-raw_max, raw_max)

cor_distfun <- function(x) as.dist(1 - cor(t(x)))

heatmaply(t(center_raw_mat),
    distfun = cor_distfun,
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    fontsize_col = 7.5,
    col = gplots::bluered(50),
    main = 'Centred log2 counts of pam50 genes',
    limits = raw_limits,
    plot_method = 'plotly')


```






```{r}
center_voom_mat <- voomed_pam50_expression - 
    apply(voomed_pam50_expression, 1, median)

voom_max <- max(abs(center_voom_mat))
voom_limits <- c(-voom_max, voom_max)



heatmaply(t(center_voom_mat),
    distfun = cor_distfun,
    row_side_colors=tcga_brca_clinical,
    fontsize_col = 7.5,
    showticklabels = c(TRUE, FALSE),
    col = gplots::bluered(50),
    limits = voom_limits,
    main = 'Normalised expression of pam50 genes',
    plot_method = 'plotly')

```

[Genomic Data Commons]: https://gdc.cancer.gov/
[data preprocessing vignette]: data_preprocessing.html
