---
title: "Using heatmaply with biological data (gene expression data)"
author: "Alan O'Callaghan"
date: "`r Sys.Date()`"
output:
  html_document:
    self_contained: yes
    toc: true
    fig_width: 15
    fig_height: 15
    depth: 3  # upto three depths of headings (specified by #, ## and ###)  
    number_sections: true  ## if you want number sections at each table header
    theme: yeti  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using heatmaply with biological data}
-->


```{r, echo = FALSE, message = FALSE}
library(heatmaply)
library(heatmaplyExamples)
library(knitr)
knitr::opts_chunk$set(
   # cache = TRUE,
   dpi = 60,
  comment = '#>',
  tidy = FALSE)

```
**Author**: Alan O'Callaghan (alan.b.ocallaghan@gmail.com)




Introduction
============

This vignette is intended to provide an overview of the use of
`heatmaply` in the analysis of biological data. In particular, 
this vignette deals with its use to visualise gene expression 
patterns and sample relationships in 
RNAseq data relating to breast cancer samples,
as retrieved from from [Genomic Data Commons].

Briefly, the following steps were performed:

1. Data retrieval pre-processing
2. Visualisation of raw and voom-transformed data (all genes)
3. Visualisation of raw and voom-transformed data 
  (median-centred data, PAM50 genes only)
4. Demonstrate outlier detection using synthetic example



Data retrieval and pre-processing
=================
The data used in this vignette is available within the package. 
A full explanation of how this data was retrieved from the 
[Genomic Data Commons] is available in the [data preprocessing vignette].


[data preprocessing vignette]: data_preprocessing.html


Visualisation of raw and voom-transformed data (all genes)
==========================================================
The R package [limma] is a commonly used statistical analysis tool using
empirical Bayes methods. This package was originally developed for micro-array data.
[voom] is a function within this package which transforms RNAseq count data in 
log2 counts per million reads, which allows it to be treated similarly. 
While pre-processing the data, quantile and TMM normalisation
were also applied. To compare these normalised values to raw read counts, 
the raw read counts were log~2~-transformed. 0.5 was
added to prevent infinite values resulting from `log2(0)`.

A common workflow for RNAseq differential expression analysis is to visualise
gene expression measures and sample-sample correlations using a heatmap. 
This is useful to observe whether samples appear to cluster together, for 
the purposes of identifying poor quality samples or outliers.
Furthermore, it may be useful to visualise differentially-expressed genes
alongside row annotations which indicate patient or sample sub-group.


```{r}
cor_mat_raw_logged <- cor(log2(raw_expression + 0.5))

heatmaply(cor_mat_raw_logged, 
    row_side_colors = tcga_brca_clinical,
    main = 'log2 Count data correlation',
    showticklabels = c(FALSE, FALSE),
    plot_method = 'plotly')

```


```{r}

cor_mat_voomed <- cor(voomed_expression)

heatmaply(cor_mat_voomed, 
    row_side_colors = tcga_brca_clinical,
    main = 'log2 cpm data correlation',
    showticklabels = c(FALSE, FALSE),
    plot_method = 'plotly')

```



Visualisation of raw data (median-centred data, PAM50 genes only)
=================================================================
Breast cancer [has been studied extensively][Breast cancer gene expression]
using gene expression profiling methods. This has lead to the identification of 
a number of gene sets which stratify patients into molecular subgroups.
One such gene set is commonly known as the [PAM50] gene set. In this example,
we will visualise gene expression patterns using heatmaply.

Centring data before performing clustering tends to result in more meaningful cluster
assignment. This is particularly true when
the measure of interest is the similarity in patterns across features, rather than 
the total distance between values. Alternatively, one could use a distance measure
which is invariant to total distance, such as correlation.
Heatmaps of non-centred are shown in 
[another vignette](non_centred_heatmaps.html) within this package. It can be seen
that the concordance between the centred and non-centred heatmaps is mediocre, and 
the clustering of triple negative samples is not as definite.

In the heatmaps shown below, it is clear that samples appear to cluster more 
clearly. Concordance with the assigned labels shown in the row annotation 
is not complete, however this may be expected, 
given that a different clustering method was used here (hierarchical clustering,
rather than [k-medioids]([Partitioning around medioids])).

```{r}
pam50_genes <- intersect(pam50_genes, rownames(raw_expression))
raw_pam50_expression <- raw_expression[pam50_genes, ]
voomed_pam50_expression <- voomed_expression[pam50_genes, ]

center_raw_mat <- cor_mat_raw_logged - 
    apply(cor_mat_raw_logged, 1, median)

raw_max <- max(abs(center_raw_mat), na.rm=TRUE)
raw_limits <- c(-raw_max, raw_max)


heatmaply(t(center_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    fontsize_col = 7.5,
    col = gplots::bluered(50),
    main = 'raw centered pam50',
    limits = raw_limits,
    plot_method = 'plotly')


heatmaply_cor(cor(center_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'correlation of raw centered pam50',
    plot_method = 'plotly')


```

Visualisation of voom-transformed data (median-centred data, PAM50 genes only)
==================
Following normalisation, gene expression patterns appear roughly similar. 
This indicates that relative expression levels have not been altered unduly.
Furthermore, slightly increased concordance with the pre-assigned cluster labels
is observed in normalised data. It may be useful when examining expression 
heatmaps to identify particularly high or low measures for a single
gene in a group of patients, or a gene which shows unusually high or low variance.
The mouse-over text available in the `heatmaply` 
package allows visual assessment of measures of interest and quick identification
of samples or genes with unusual gene expression patterns.
Similarly, visualising correlation heatmaps with `heatmaply` allows the user to
rapidly identify samples with unusually high or low pairwise correlation.


```{r}
center_voom_mat <- voomed_pam50_expression - 
    apply(voomed_pam50_expression, 1, median)

voom_max <- max(abs(center_voom_mat))
voom_limits <- c(-voom_max, voom_max)


heatmaply(t(center_voom_mat), 
    row_side_colors=tcga_brca_clinical,
    fontsize_col = 7.5,
    showticklabels = c(TRUE, FALSE),
    col = gplots::bluered(50),
    limits = voom_limits,
    main = 'voomed pam50',
    plot_method = 'plotly')


heatmaply_cor(cor(center_voom_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'correlation of voomed pam50',
    plot_method = 'plotly')

```

References
==========
- [limma]
- [voom](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)
- [Trimmed mean of M-component normalisation][TMM]
- [Genomics Data Commons]
- [PAM50]
- [Breast cancer gene expression](https://www.ncbi.nlm.nih.gov/pubmed/28733194)
- [Partitioning around medioids]

[Partitioning around medioids]: https://en.wikipedia.org/wiki/K-medoids
[Genomics Data Commons]: https://gdc.cancer.gov/
[limma]: https://bioconductor.org/packages/release/bioc/html/limma.html
[TMM]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2864565/
[PAM50]: http://ascopubs.org/doi/abs/10.1200/jco.2008.18.1370
