---
title: "Using heatmaply with biological data (mutational and gene expression data)"
date: "`r Sys.Date()`"
author: "Alan O'Callaghan"
output:
  html_document:
    self_contained: yes
    toc: true
    fig_width: 15
    fig_height: 15
    depth: 3  # upto three depths of headings (specified by #, ## and ###)  
    number_sections: true  ## if you want number sections at each table header
    theme: yeti  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using heatmaply with biological data}
-->


```{r, echo = FALSE, message = FALSE}
library(heatmaply)
library(heatmaplyExamples)
library(knitr)
knitr::opts_chunk$set(
   # cache = TRUE,
   dpi = 60,
  comment = '#>',
  tidy = FALSE)

```
**Author**: Alan O'Callaghan (alan.b.ocallaghan@gmail.com)




Introduction
============

- Data pre-processing
- See if samples cluster based on receptor status
- Subset to PAM50 genes
- See if samples cluster based on receptor status
- Center data
- Compare voom and normal log2
- Demonstrate outlier detection


Data pre-processing
=================
The data used in this vignette is available within the package. A full 
explanation of how this data was retrieved from the 
[Genomic Data Commons] is available in the [data preprocessing vignette].


[Genomics Data Commons]: https://gdc.cancer.gov/
[data preprocessing vignette]: ./data_preprocessing.html


heatmaply workflow
==================
```{r}
cor_mat_raw_logged <- cor(log2(raw_expression + 0.5))

heatmaply(cor_mat_raw_logged, 
    main = 'log2 Count data correlation',
    showticklabels = c(FALSE, FALSE),
    row_side_colors=tcga_brca_clinical,
    plot_method = 'plotly')

```
voom explanation

```{r}

cor_mat_voomed <- cor(voomed_expression)

heatmaply(cor_mat_voomed, 
    main = 'log2 cpm data correlation',
    showticklabels = c(FALSE, FALSE),
    row_side_colors=tcga_brca_clinical,
    plot_method = 'plotly')


tmp <- voomed_expression
tmp[, 1] <- tmp[, 1] <- rnorm(nrow(tmp), 100, 20)


heatmaply(cor(voomed_expression), 
    row_side_colors=tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'Outlier on correlation matrix',
    col = gplots::bluered(50),
    limits = c(-1, 1),
    plot_method = 'plotly')


```



```{r}
pam50_genes <- intersect(pam50_genes, rownames(raw_expression))
raw_pam50_expression <- raw_expression[pam50_genes, ]
voomed_pam50_expression <- voomed_expression[pam50_genes, ]

center_raw_mat <- log2(raw_pam50_expression + 0.5) - 
    apply(log2(raw_pam50_expression + 0.5), 1, median)

raw_max <- max(abs(center_raw_mat), na.rm=TRUE)
raw_limits <- c(-raw_max, raw_max)


heatmaply(t(center_raw_mat), 
    row_side_colors=tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    col = gplots::bluered(50),
    main = 'raw centered pam50',
    limits = raw_limits,
    plot_method = 'plotly')


heatmaply(cor(center_raw_mat), 
    row_side_colors=tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'correlation of raw centered pam50',
    limits = c(-1, 1),
    col = gplots::bluered(50),
    plot_method = 'plotly')


center_voom_mat <- voomed_pam50_expression - 
    apply(voomed_pam50_expression, 1, median)

voom_max <- max(abs(center_voom_mat))
voom_limits <- c(-voom_max, voom_max)


heatmaply(t(center_voom_mat), 
    row_side_colors=tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    col = gplots::bluered(50),
    limits = voom_limits,
    main = 'voomed pam50',
    plot_method = 'plotly')


heatmaply(cor(center_voom_mat), 
    row_side_colors=tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'correlation of voomed pam50',
    col = gplots::bluered(50),
    limits = c(-1, 1),
    plot_method = 'plotly')



```
