---
title: "Using heatmaply with non-centred RNAseq heatmaps (PAM50 genes)"
author: "Alan O'Callaghan"
date: "`r Sys.Date()`"
output:
  html_document:
    self_contained: yes
    toc: true
    fig_width: 14
    fig_height: 10
    depth: 3  # upto three depths of headings (specified by #, ## and ###)  
    number_sections: true  ## if you want number sections at each table header
    theme: yeti  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using heatmaply with non-centred RNAseq heatmaps (PAM50 genes)}
-->


```{r, echo = FALSE, message = FALSE}
library(heatmaply)
library(heatmaplyExamples)
library(knitr)
knitr::opts_chunk$set(
   # cache = TRUE,
   dpi = 60,
  comment = '#>',
  tidy = FALSE)

```
**Author**: Alan O'Callaghan (alan.b.ocallaghan@gmail.com)

Introduction
============
This vignette illustrates the clustering of non-centered breast
cancer RNAseq data, similar to the centered data. shown 
in the main [biological data](biological_data.html)
vignette in this package.


```{r}

pam50_genes <- intersect(pam50_genes, rownames(raw_expression))
raw_pam50_expression <- raw_expression[pam50_genes, ]
voomed_pam50_expression <- voomed_expression[pam50_genes, ]

log_raw_mat <- log2(raw_pam50_expression + 0.5)

heatmaply(t(log_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    fontsize_col = 7.5,
    col = gplots::bluered(50),
    main = 'Pre-normalisation log2 counts, PAM50 genes',
    plot_method = 'plotly')



```


```{r, fig.width=13, fig.height=10}


heatmaply_cor(cor(log_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'Sample-sample correlation based on log2-transformed PAM50 gene expression',
    plot_method = 'plotly')



```

```{r}



heatmaply(t(voomed_pam50_expression), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    fontsize_col = 7.5,
    col = gplots::bluered(50),
    main = 'Normalised log2 CPM, PAM50 genes',
    plot_method = 'plotly')

```

```{r, fig.width=13, fig.height=10}


heatmaply_cor(cor(voomed_pam50_expression), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'Sample-sample correlation based on normalised PAM50 gene expression',
    plot_method = 'plotly')

```


sessionInfo
===============


```{r}
sessionInfo()
```

