---
title: "Non-centred RNAseq heatmap (PAM50 genes)"
date: "`r Sys.Date()`"
author: "Alan O'Callaghan"
output:
  html_document:
    self_contained: yes
    toc: true
    fig_width: 15
    fig_height: 15
    depth: 3  # upto three depths of headings (specified by #, ## and ###)  
    number_sections: true  ## if you want number sections at each table header
    theme: yeti  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Using heatmaply with biological data}
-->


```{r, echo = FALSE, message = FALSE}
library(heatmaply)
library(heatmaplyExamples)
library(knitr)
knitr::opts_chunk$set(
   # cache = TRUE,
   dpi = 60,
  comment = '#>',
  tidy = FALSE)

```
**Author**: Alan O'Callaghan (alan.b.ocallaghan@gmail.com)

Introduction
============
This vignette illustrates the clustering of non-centred breast
cancer RNAseq data, similar to the centred data. shown 
in the main [biological data](biological_data.html)
vignette in this package.



Non-centred heatmaps
============================





```{r}

pam50_genes <- intersect(pam50_genes, rownames(tcga_breast_expression))
raw_pam50_expression <- tcga_breast_expression[pam50_genes, ]


dge <- edgeR::DGEList(counts = tcga_breast_expression)
dge <- edgeR::calcNormFactors(dge, method = "TMM")
voomed_tcga_expression <- limma::voom(dge, normalize.method="quantile")
voomed_tcga_expression <- voomed_tcga_expression$E

voomed_pam50_expression <- voomed_tcga_expression[pam50_genes, ]

cor_distfun <- function(x) as.dist(1 - cor(t(x)))


log_raw_mat <- log2(raw_pam50_expression + 0.5)



heatmaply(t(log_raw_mat),
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    fontsize_col = 7.5,
    distfun = cor_distfun,
    col = gplots::bluered(50),
    main = 'log2 expression of count data (pam50 genes)',
    plot_method = 'plotly')


heatmaply_cor(cor(log_raw_mat), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'Sample correlation of raw count data (pam50 genes)',
    plot_method = 'plotly')
```






```{r}


heatmaply(t(voomed_pam50_expression), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(TRUE, FALSE),
    fontsize_col = 7.5,
    distfun = cor_distfun,
    col = gplots::bluered(50),
    main = 'Normalised RNAseq gene expression data (pam50 genes)',
    plot_method = 'plotly')


heatmaply_cor(cor(voomed_pam50_expression), 
    row_side_colors = tcga_brca_clinical,
    showticklabels = c(FALSE, FALSE),
    main = 'Sample correlation of normalised pam50 genes',
    plot_method = 'plotly')


```
